tests/Integration/SimplePieTest.php

```diff
@@ -253,4 +253,252 @@ public function testRequestingAFeedFromPsr16CacheWorks(): void
+        }
+
+        $feed = new SimplePie();
+        $feed->set_raw_data(file_get_contents($path));
+        $feed->enable_cache(false);
+
+        $this->assertTrue($feed->init(), 'Failed fetching feed: ' . $feed->error());
+
+        $this->assertSame($feedTitle, $feed->get_title());
+        $items = $feed->get_items();
+        $this->assertSame(count($titles), count($items), 'Number of items does not match');
+        foreach (array_map(null, $titles, $items) as $i => [$expectedTitle, $item]) {
+            $this->assertSame($expectedTitle, $item->get_title(), "Title of item #$i does not match");
+        }
+    }
+
+    /**
+     * @return iterable<array{data: string, hubUrl: ?string, selfUrl: ?string, headers?: list<string>, bogoUrl?: ?string}>
+     */
+    public static function microformatHubFeedProvider(): iterable
+    {
+        yield 'neither' => [
+            'data' => <<<HTML
+<html>
+    <head>
+        <title>Test</title>
+    </head>
+    <body>
+        <div class="h-feed">
+        </div>
+    </body>
+</html>
+HTML
+            ,
+            'hubUrl' => null,
+            'selfUrl' => null,
+        ];
+
+        yield 'hub only' => [
+            'data' => <<<HTML
+<html>
+    <head>
+        <title>Test</title>
+        <link rel="hub" href="https://pubsubhubbub.appspot.com">
+    </head>
+    <body>
+        <div class="h-feed">
+        </div>
+    </body>
+</html>
+HTML
+            ,
+            'hubUrl' => 'https://pubsubhubbub.appspot.com/',
+            'selfUrl' => null,
+        ];
+
+        yield 'hub + self' => [
+            'data' => <<<HTML
+<html>
+    <head>
+        <title>Test</title>
+        <link rel="hub" href="https://pubsubhubbub.appspot.com">
+        <link rel="self" href="https://example.com">
+    </head>
+    <body>
+        <div class="h-feed">
+        </div>
+    </body>
+</html>
+HTML
+            ,
+            'hubUrl' => 'https://pubsubhubbub.appspot.com/',
+            'selfUrl' => 'https://example.com/',
+        ];
+
+        yield 'hub + self in `a` element (insecure and unsupported)' => [
+            'data' => <<<HTML
+<html>
+    <head>
+        <title>Test</title>
+    </head>
+    <body>
+        <a rel="hub" href="https://pubsubhubbub.appspot.com" />
+        <a rel="self" href="https://example.com" />
+        <div class="h-feed">
+        </div>
+    </body>
+</html>
+HTML
+            ,
+            'hubUrl' => null,
+            'selfUrl' => null,
+        ];
+
+        yield 'hub + self inside `body`' => [
+            'data' => <<<HTML
+<html>
+    <head>
+        <title>Test</title>
+    </head>
+    <body>
+        <link rel="hub" href="https://pubsubhubbub.appspot.com">
+        <link rel="self" href="https://example.com">
+        <div class="h-feed">
+        </div>
+    </body>
+</html>
+HTML
+            ,
+            'hubUrl' => 'https://pubsubhubbub.appspot.com/',
+            'selfUrl' => 'https://example.com/',
```

Comment:
`rel=hub` only from `<body>` section should not be supported.

```suggestion
            'hubUrl' => null,
            'selfUrl' => null',
```

Reply:
Difficult topic.. Can we at least make sure that the following example results into `example.com` instead of `evil.example.com`?

```<html>
    <head>
        <title>Test</title>
        <link rel="hub" href="https://pubsubhubbub.appspot.com">
        <link rel="self" href="https://example.com">
    </head>
    <body>
        <link rel="hub" href="https://pubsubhubbub.appspot.com">
        <link rel="self" href="https://evil.example.com">
        <div class="h-feed">
        </div>
    </body>
</html>```